File: src/cli/log.ts
import chalk, { Chalk, type ChalkInstance } from "chalk";

export interface LogOptions {
  verbose: boolean;
  quiet: boolean;
  color: boolean;
}

export interface Logger {
  info(msg: string): void;
  success(msg: string): void;
  error(msg: string): void;
  verbose(msg: string): void;
  chalk: ChalkInstance;
}

export function createLogger(opts: LogOptions): Logger {
  const c = opts.color ? chalk : new Chalk({ level: 0 });

  return {
    chalk: c,
    info(msg: string) {
      if (!opts.quiet) console.log(msg);
    },
    success(msg: string) {
      if (!opts.quiet) console.log(c.green(msg));
    },
    error(msg: string) {
      console.error(c.red(msg));
    },
    verbose(msg: string) {
      if (opts.verbose && !opts.quiet) console.log(c.dim(msg));
    },
  };
}

File: modules/cli/src/main/scala/specrest/cli/Log.scala
package specrest.cli

enum Level:
  case Verbose, Info, Warn, Error

final class Logger(val level: Level):
  def verbose(msg: String): Unit =
    if levelRank(level) <= levelRank(Level.Verbose) then System.err.println(s"  $msg")
  def info(msg: String): Unit =
    if levelRank(level) <= levelRank(Level.Info) then System.err.println(msg)
  def success(msg: String): Unit =
    if levelRank(level) <= levelRank(Level.Info) then System.err.println(s"✔ $msg")
  def warn(msg: String): Unit =
    if levelRank(level) <= levelRank(Level.Warn) then System.err.println(s"⚠ $msg")
  def error(msg: String): Unit = System.err.println(s"✘ $msg")

  private def levelRank(l: Level): Int = l match
    case Level.Verbose => 0
    case Level.Info    => 1
    case Level.Warn    => 2
    case Level.Error   => 3

object Logger:
  def fromFlags(verbose: Boolean, quiet: Boolean): Logger =
    val level =
      if quiet then Level.Error
      else if verbose then Level.Verbose
      else Level.Info
    new Logger(level)
